### PR TITLE
Storage Blob LeaseClient

### PR DESCRIPTION
**Pull request by alzimmermsft**
Implements #5379 

Adds LeaseClient and LeaseAsyncClient to the specialized package and removes the leasing methods from BlobClient, BlobAsyncClient, ContainerClient, and ContainerAsyncClient.

Updates tests and docs based on the new code changes.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+                .map(rb -> new SimpleResponse<>(rb, rb.getDeserializedHeaders().getLeaseTime()));
+        } else {
+            return postProcessResponse(this.client.containers().breakLeaseWithRestResponseAsync(
+                null, null, breakPeriodInSeconds, null, modifiedAccessConditions, context))
+                .map(rb -> new SimpleResponse<>(rb, rb.getDeserializedHeaders().getLeaseTime()));
+        }
+    }
+
+    /**
+     * ChangeLease changes the blob's lease ID.
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java]:
```suggestion
     * Changes the blob's lease ID.
```

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+     *
+     * @param proposedId A {@code String} in any valid GUID format.
+     * @return A reactive response containing the new lease ID.
+     */
+    public Mono<String> changeLease(String proposedId) {
+        return changeLeaseWithResponse(proposedId, null).flatMap(FluxUtil::toMono);
+    }
+
+    /**
+     * ChangeLease changes the blob's lease ID.
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java]:
```suggestion
     * Changes the blob's lease ID.
```

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClient.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+import com.azure.storage.blob.ContainerClient;
+import com.azure.storage.blob.models.ModifiedAccessConditions;
+import com.azure.storage.common.Utility;
+
+import java.net.URL;
+import java.time.Duration;
+
+/**
+ *
+ */
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClient.java]: JavaDoc

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClient.java [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        this.client = client;
+    }
+
+    /**
+     * Gets the {@link URL} of the lease client. If the lease client is using a {@link BlobClient} it will be the URL of
+     * the blob, otherwise the URL will be the {@link ContainerClient} URL.
+     *
+     * @return URL of the lease client.
+     */
+    public URL getLeaseClientUrl() {
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClient.java]: Apply same comments from the async client here also.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClientBuilder.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+import java.net.URL;
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * This class provides a fluent builder API to help aid the configuration and instantiation of Storage Blob lease
+ * clients. A new instance of {@link LeaseClient} and {@link LeaseAsyncClient} are constructed every time
+ * {@link #buildClient() buildClient} and {@link #buildAsyncClient() buildAsyncClient} are called
+ * respectively.
```
Comment (JonathanGiles) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClientBuilder.java]: This JavaDoc is blob-specific. You should also note the container support.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+            .build();
+    }
+
+    /**
+     * Gets the {@link URL} of the lease client. If the lease client is using a {@link BlobAsyncClient} it will be the
+     * URL of the blob, otherwise the URL will be the {@link ContainerAsyncClient} URL.
+     *
+     * @return URL of the lease client.
+     */
+    public URL getLeaseAsyncClientUrl() {
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java]: Changed to `getLeaseUrl`

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+     * {@codesnippet com.azure.storage.blob.specialized.LeaseAsyncClient.acquireLease#int}
+     *
+     * <p>For more information, see the
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/lease-blob">Azure Docs</a></p>
+     *
+     * @param duration The  duration of the lease, in seconds, or negative one (-1) for a lease that never expires. A
+     * non-infinite lease can be between 15 and 60 seconds.
+     * @return A reactive response containing the lease ID.
+     */
+    public Mono<String> acquireLease(int duration) {
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java]: `Duration` has no concept of infinite and having `null` indicating infinite is strange, rather leave it as is.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClient.java (hunk 4 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseClient.java]: `getLeaseUrl`

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        } else {
+            return postProcessResponse(this.client.containers().releaseLeaseWithRestResponseAsync(
+                null, this.leaseId, null, null, modifiedAccessConditions, context))
+                .map(VoidResponse::new);
+        }
+    }
+
+    /**
+     * BreakLease breaks the blob's previously-acquired lease (if it exists). Pass the LeaseBreakDefault (-1) constant
+     * to break a fixed-duration lease when it expires or an infinite lease immediately.
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/LeaseAsyncClient.java]: Removed for now as the REST docs seem to indicate `null` is the equivalent for break now.

**Pull request merged**